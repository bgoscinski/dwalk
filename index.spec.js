const dwalk = require('./index.js')
const FIXTURE = './fixture'

const canonPath = (p) => p.replace(/\\/g, '/')

it('should calculate relative path', () => {
  const files = [...dwalk(FIXTURE)]
  const clean = files.map((d) => `${d.depth} ${canonPath(d.relative)}`)
  expect(clean).toMatchSnapshot()
})

it('should have name property', () => {
  const all = [...dwalk(FIXTURE)]
  const clean = all.map((d) => `${d.depth} ${d.name}`)
  expect(clean).toMatchSnapshot()
})

it('should calculate absolute path', () => {
  const all = [...dwalk(FIXTURE)]
  const clean = all.map((d) => `${d.depth} ${canonPath(d.absolute)}`)
  expect(clean).toMatchSnapshot()
})

it('should handle skip command', () => {
  const gen = dwalk(FIXTURE)
  expect(gen.next().value).toMatchObject({ relative: '.gitkeep' })

  expect(gen.next().value).toMatchObject({ relative: 'level-1-a' })
  // signal it to skip contents of level-1-a
  expect(gen.next('skip').value).toMatchObject({ relative: 'level-1-b' })

  expect(gen.next().value).toMatchObject({ relative: 'level-1-c' })

  // without skipping it would be level-1-a/.gitkeep
  expect(gen.next().value).toMatchObject({ relative: 'level-1-b\\.gitkeep' })
})

it('should handle stop command', () => {
  const gen = dwalk(FIXTURE)
  expect(gen.next().value).toMatchObject({ relative: '.gitkeep' })

  expect(gen.next('stop').value).toBe(undefined)
  expect(gen.next().value).toBe(undefined)
})

it('should handle false as stop command', () => {
  const gen = dwalk(FIXTURE)
  expect(gen.next().value).toMatchObject({ relative: '.gitkeep' })

  expect(gen.next(false).value).toBe(undefined)
  expect(gen.next().value).toBe(undefined)
})

describe('.cb', () => {
  it('should invoke callback with same files as normal', () => {
    const files = []
    dwalk.cb(FIXTURE, (ev) => files.push(ev))
    expect(files).toEqual([...dwalk(FIXTURE)])
  })

  it('should handle skip command', () => {
    const files = []
    dwalk.cb(FIXTURE, (file) => {
      if (file.relative === 'level-1-a') return 'skip'
      files.push(file)
    })

    expect(files.length).toBe(53)
    const anyLevel1A = files.filter((ev) => ev.relative.includes('level-1-a'))
    expect(anyLevel1A).toEqual([])
  })

  it('should handle stop command', () => {
    const files = []
    dwalk.cb(FIXTURE, (file) => {
      if (file.relative === 'level-1-a') return 'stop'
      files.push(file)
    })
    expect(files.length).toBe(1)
  })

  it('should handle false as stop command', () => {
    const files = []
    dwalk.cb(FIXTURE, (file) => {
      if (file.relative === 'level-1-a') return false
      files.push(file)
    })
    expect(files.length).toBe(1)
  })

  it('should handle skip_dir command', () => {
    let found = false
    const pre = []
    const post = []

    dwalk.cb(FIXTURE, (file) => {
      if (found) return post.push(file)

      pre.push(file)

      if ((found = file.relative.endsWith('level-2-b'))) {
        return 'skip_dir'
      }
    })

    expect(pre.map((f) => canonPath(f.relative))).toMatchSnapshot()
    expect(post.map((f) => canonPath(f.relative))).toMatchSnapshot()
  })
})

it('should not emit anything when passed path is file', () => {
  const all = [...dwalk('package.json')]
  expect(all).toEqual([])
})
