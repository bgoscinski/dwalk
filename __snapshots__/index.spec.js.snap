// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.cb should handle skip_dir command 1`] = `
Array [
  ".gitkeep",
  "level-1-a",
  "level-1-b",
  "level-1-c",
  "level-1-a/.gitkeep",
  "level-1-a/level-2-a",
  "level-1-a/level-2-b",
]
`;

exports[`.cb should handle skip_dir command 2`] = `
Array [
  "level-1-b/.gitkeep",
  "level-1-b/level-2-a",
  "level-1-b/level-2-b",
  "level-1-b/level-2-c",
  "level-1-c/.gitkeep",
  "level-1-c/level-2-a",
  "level-1-c/level-2-b",
  "level-1-c/level-2-c",
  "level-1-a/level-2-a/.gitkeep",
  "level-1-a/level-2-a/level-3-a",
  "level-1-a/level-2-a/level-3-b",
  "level-1-a/level-2-a/level-3-c",
  "level-1-b/level-2-a/.gitkeep",
  "level-1-b/level-2-a/level-3-a",
  "level-1-b/level-2-a/level-3-b",
  "level-1-b/level-2-a/level-3-c",
  "level-1-b/level-2-b/.gitkeep",
  "level-1-b/level-2-b/level-3-a",
  "level-1-b/level-2-b/level-3-b",
  "level-1-b/level-2-b/level-3-c",
  "level-1-b/level-2-c/.gitkeep",
  "level-1-b/level-2-c/level-3-a",
  "level-1-b/level-2-c/level-3-b",
  "level-1-b/level-2-c/level-3-c",
  "level-1-c/level-2-a/.gitkeep",
  "level-1-c/level-2-a/level-3-a",
  "level-1-c/level-2-a/level-3-b",
  "level-1-c/level-2-a/level-3-c",
  "level-1-c/level-2-b/.gitkeep",
  "level-1-c/level-2-b/level-3-a",
  "level-1-c/level-2-b/level-3-b",
  "level-1-c/level-2-b/level-3-c",
  "level-1-c/level-2-c/.gitkeep",
  "level-1-c/level-2-c/level-3-a",
  "level-1-c/level-2-c/level-3-b",
  "level-1-c/level-2-c/level-3-c",
  "level-1-a/level-2-a/level-3-a/.gitkeep",
  "level-1-a/level-2-a/level-3-b/.gitkeep",
  "level-1-a/level-2-a/level-3-c/.gitkeep",
  "level-1-b/level-2-a/level-3-a/.gitkeep",
  "level-1-b/level-2-a/level-3-b/.gitkeep",
  "level-1-b/level-2-a/level-3-c/.gitkeep",
  "level-1-b/level-2-b/level-3-a/.gitkeep",
  "level-1-b/level-2-b/level-3-b/.gitkeep",
  "level-1-b/level-2-b/level-3-c/.gitkeep",
  "level-1-b/level-2-c/level-3-a/.gitkeep",
  "level-1-b/level-2-c/level-3-b/.gitkeep",
  "level-1-b/level-2-c/level-3-c/.gitkeep",
  "level-1-c/level-2-a/level-3-a/.gitkeep",
  "level-1-c/level-2-a/level-3-b/.gitkeep",
  "level-1-c/level-2-a/level-3-c/.gitkeep",
  "level-1-c/level-2-b/level-3-a/.gitkeep",
  "level-1-c/level-2-b/level-3-b/.gitkeep",
  "level-1-c/level-2-b/level-3-c/.gitkeep",
  "level-1-c/level-2-c/level-3-a/.gitkeep",
  "level-1-c/level-2-c/level-3-b/.gitkeep",
  "level-1-c/level-2-c/level-3-c/.gitkeep",
]
`;

exports[`should calculate absolute path 1`] = `
Array [
  "1 C:/dev/dwalk/fixture/.gitkeep",
  "1 C:/dev/dwalk/fixture/level-1-a",
  "1 C:/dev/dwalk/fixture/level-1-b",
  "1 C:/dev/dwalk/fixture/level-1-c",
  "2 C:/dev/dwalk/fixture/level-1-a/.gitkeep",
  "2 C:/dev/dwalk/fixture/level-1-a/level-2-a",
  "2 C:/dev/dwalk/fixture/level-1-a/level-2-b",
  "2 C:/dev/dwalk/fixture/level-1-a/level-2-c",
  "2 C:/dev/dwalk/fixture/level-1-b/.gitkeep",
  "2 C:/dev/dwalk/fixture/level-1-b/level-2-a",
  "2 C:/dev/dwalk/fixture/level-1-b/level-2-b",
  "2 C:/dev/dwalk/fixture/level-1-b/level-2-c",
  "2 C:/dev/dwalk/fixture/level-1-c/.gitkeep",
  "2 C:/dev/dwalk/fixture/level-1-c/level-2-a",
  "2 C:/dev/dwalk/fixture/level-1-c/level-2-b",
  "2 C:/dev/dwalk/fixture/level-1-c/level-2-c",
  "3 C:/dev/dwalk/fixture/level-1-a/level-2-a/.gitkeep",
  "3 C:/dev/dwalk/fixture/level-1-a/level-2-a/level-3-a",
  "3 C:/dev/dwalk/fixture/level-1-a/level-2-a/level-3-b",
  "3 C:/dev/dwalk/fixture/level-1-a/level-2-a/level-3-c",
  "3 C:/dev/dwalk/fixture/level-1-a/level-2-b/.gitkeep",
  "3 C:/dev/dwalk/fixture/level-1-a/level-2-b/level-3-a",
  "3 C:/dev/dwalk/fixture/level-1-a/level-2-b/level-3-b",
  "3 C:/dev/dwalk/fixture/level-1-a/level-2-b/level-3-c",
  "3 C:/dev/dwalk/fixture/level-1-a/level-2-c/.gitkeep",
  "3 C:/dev/dwalk/fixture/level-1-a/level-2-c/level-3-a",
  "3 C:/dev/dwalk/fixture/level-1-a/level-2-c/level-3-b",
  "3 C:/dev/dwalk/fixture/level-1-a/level-2-c/level-3-c",
  "3 C:/dev/dwalk/fixture/level-1-b/level-2-a/.gitkeep",
  "3 C:/dev/dwalk/fixture/level-1-b/level-2-a/level-3-a",
  "3 C:/dev/dwalk/fixture/level-1-b/level-2-a/level-3-b",
  "3 C:/dev/dwalk/fixture/level-1-b/level-2-a/level-3-c",
  "3 C:/dev/dwalk/fixture/level-1-b/level-2-b/.gitkeep",
  "3 C:/dev/dwalk/fixture/level-1-b/level-2-b/level-3-a",
  "3 C:/dev/dwalk/fixture/level-1-b/level-2-b/level-3-b",
  "3 C:/dev/dwalk/fixture/level-1-b/level-2-b/level-3-c",
  "3 C:/dev/dwalk/fixture/level-1-b/level-2-c/.gitkeep",
  "3 C:/dev/dwalk/fixture/level-1-b/level-2-c/level-3-a",
  "3 C:/dev/dwalk/fixture/level-1-b/level-2-c/level-3-b",
  "3 C:/dev/dwalk/fixture/level-1-b/level-2-c/level-3-c",
  "3 C:/dev/dwalk/fixture/level-1-c/level-2-a/.gitkeep",
  "3 C:/dev/dwalk/fixture/level-1-c/level-2-a/level-3-a",
  "3 C:/dev/dwalk/fixture/level-1-c/level-2-a/level-3-b",
  "3 C:/dev/dwalk/fixture/level-1-c/level-2-a/level-3-c",
  "3 C:/dev/dwalk/fixture/level-1-c/level-2-b/.gitkeep",
  "3 C:/dev/dwalk/fixture/level-1-c/level-2-b/level-3-a",
  "3 C:/dev/dwalk/fixture/level-1-c/level-2-b/level-3-b",
  "3 C:/dev/dwalk/fixture/level-1-c/level-2-b/level-3-c",
  "3 C:/dev/dwalk/fixture/level-1-c/level-2-c/.gitkeep",
  "3 C:/dev/dwalk/fixture/level-1-c/level-2-c/level-3-a",
  "3 C:/dev/dwalk/fixture/level-1-c/level-2-c/level-3-b",
  "3 C:/dev/dwalk/fixture/level-1-c/level-2-c/level-3-c",
  "4 C:/dev/dwalk/fixture/level-1-a/level-2-a/level-3-a/.gitkeep",
  "4 C:/dev/dwalk/fixture/level-1-a/level-2-a/level-3-b/.gitkeep",
  "4 C:/dev/dwalk/fixture/level-1-a/level-2-a/level-3-c/.gitkeep",
  "4 C:/dev/dwalk/fixture/level-1-a/level-2-b/level-3-a/.gitkeep",
  "4 C:/dev/dwalk/fixture/level-1-a/level-2-b/level-3-b/.gitkeep",
  "4 C:/dev/dwalk/fixture/level-1-a/level-2-b/level-3-c/.gitkeep",
  "4 C:/dev/dwalk/fixture/level-1-a/level-2-c/level-3-a/.gitkeep",
  "4 C:/dev/dwalk/fixture/level-1-a/level-2-c/level-3-b/.gitkeep",
  "4 C:/dev/dwalk/fixture/level-1-a/level-2-c/level-3-c/.gitkeep",
  "4 C:/dev/dwalk/fixture/level-1-b/level-2-a/level-3-a/.gitkeep",
  "4 C:/dev/dwalk/fixture/level-1-b/level-2-a/level-3-b/.gitkeep",
  "4 C:/dev/dwalk/fixture/level-1-b/level-2-a/level-3-c/.gitkeep",
  "4 C:/dev/dwalk/fixture/level-1-b/level-2-b/level-3-a/.gitkeep",
  "4 C:/dev/dwalk/fixture/level-1-b/level-2-b/level-3-b/.gitkeep",
  "4 C:/dev/dwalk/fixture/level-1-b/level-2-b/level-3-c/.gitkeep",
  "4 C:/dev/dwalk/fixture/level-1-b/level-2-c/level-3-a/.gitkeep",
  "4 C:/dev/dwalk/fixture/level-1-b/level-2-c/level-3-b/.gitkeep",
  "4 C:/dev/dwalk/fixture/level-1-b/level-2-c/level-3-c/.gitkeep",
  "4 C:/dev/dwalk/fixture/level-1-c/level-2-a/level-3-a/.gitkeep",
  "4 C:/dev/dwalk/fixture/level-1-c/level-2-a/level-3-b/.gitkeep",
  "4 C:/dev/dwalk/fixture/level-1-c/level-2-a/level-3-c/.gitkeep",
  "4 C:/dev/dwalk/fixture/level-1-c/level-2-b/level-3-a/.gitkeep",
  "4 C:/dev/dwalk/fixture/level-1-c/level-2-b/level-3-b/.gitkeep",
  "4 C:/dev/dwalk/fixture/level-1-c/level-2-b/level-3-c/.gitkeep",
  "4 C:/dev/dwalk/fixture/level-1-c/level-2-c/level-3-a/.gitkeep",
  "4 C:/dev/dwalk/fixture/level-1-c/level-2-c/level-3-b/.gitkeep",
  "4 C:/dev/dwalk/fixture/level-1-c/level-2-c/level-3-c/.gitkeep",
]
`;

exports[`should calculate relative path 1`] = `
Array [
  "1 .gitkeep",
  "1 level-1-a",
  "1 level-1-b",
  "1 level-1-c",
  "2 level-1-a/.gitkeep",
  "2 level-1-a/level-2-a",
  "2 level-1-a/level-2-b",
  "2 level-1-a/level-2-c",
  "2 level-1-b/.gitkeep",
  "2 level-1-b/level-2-a",
  "2 level-1-b/level-2-b",
  "2 level-1-b/level-2-c",
  "2 level-1-c/.gitkeep",
  "2 level-1-c/level-2-a",
  "2 level-1-c/level-2-b",
  "2 level-1-c/level-2-c",
  "3 level-1-a/level-2-a/.gitkeep",
  "3 level-1-a/level-2-a/level-3-a",
  "3 level-1-a/level-2-a/level-3-b",
  "3 level-1-a/level-2-a/level-3-c",
  "3 level-1-a/level-2-b/.gitkeep",
  "3 level-1-a/level-2-b/level-3-a",
  "3 level-1-a/level-2-b/level-3-b",
  "3 level-1-a/level-2-b/level-3-c",
  "3 level-1-a/level-2-c/.gitkeep",
  "3 level-1-a/level-2-c/level-3-a",
  "3 level-1-a/level-2-c/level-3-b",
  "3 level-1-a/level-2-c/level-3-c",
  "3 level-1-b/level-2-a/.gitkeep",
  "3 level-1-b/level-2-a/level-3-a",
  "3 level-1-b/level-2-a/level-3-b",
  "3 level-1-b/level-2-a/level-3-c",
  "3 level-1-b/level-2-b/.gitkeep",
  "3 level-1-b/level-2-b/level-3-a",
  "3 level-1-b/level-2-b/level-3-b",
  "3 level-1-b/level-2-b/level-3-c",
  "3 level-1-b/level-2-c/.gitkeep",
  "3 level-1-b/level-2-c/level-3-a",
  "3 level-1-b/level-2-c/level-3-b",
  "3 level-1-b/level-2-c/level-3-c",
  "3 level-1-c/level-2-a/.gitkeep",
  "3 level-1-c/level-2-a/level-3-a",
  "3 level-1-c/level-2-a/level-3-b",
  "3 level-1-c/level-2-a/level-3-c",
  "3 level-1-c/level-2-b/.gitkeep",
  "3 level-1-c/level-2-b/level-3-a",
  "3 level-1-c/level-2-b/level-3-b",
  "3 level-1-c/level-2-b/level-3-c",
  "3 level-1-c/level-2-c/.gitkeep",
  "3 level-1-c/level-2-c/level-3-a",
  "3 level-1-c/level-2-c/level-3-b",
  "3 level-1-c/level-2-c/level-3-c",
  "4 level-1-a/level-2-a/level-3-a/.gitkeep",
  "4 level-1-a/level-2-a/level-3-b/.gitkeep",
  "4 level-1-a/level-2-a/level-3-c/.gitkeep",
  "4 level-1-a/level-2-b/level-3-a/.gitkeep",
  "4 level-1-a/level-2-b/level-3-b/.gitkeep",
  "4 level-1-a/level-2-b/level-3-c/.gitkeep",
  "4 level-1-a/level-2-c/level-3-a/.gitkeep",
  "4 level-1-a/level-2-c/level-3-b/.gitkeep",
  "4 level-1-a/level-2-c/level-3-c/.gitkeep",
  "4 level-1-b/level-2-a/level-3-a/.gitkeep",
  "4 level-1-b/level-2-a/level-3-b/.gitkeep",
  "4 level-1-b/level-2-a/level-3-c/.gitkeep",
  "4 level-1-b/level-2-b/level-3-a/.gitkeep",
  "4 level-1-b/level-2-b/level-3-b/.gitkeep",
  "4 level-1-b/level-2-b/level-3-c/.gitkeep",
  "4 level-1-b/level-2-c/level-3-a/.gitkeep",
  "4 level-1-b/level-2-c/level-3-b/.gitkeep",
  "4 level-1-b/level-2-c/level-3-c/.gitkeep",
  "4 level-1-c/level-2-a/level-3-a/.gitkeep",
  "4 level-1-c/level-2-a/level-3-b/.gitkeep",
  "4 level-1-c/level-2-a/level-3-c/.gitkeep",
  "4 level-1-c/level-2-b/level-3-a/.gitkeep",
  "4 level-1-c/level-2-b/level-3-b/.gitkeep",
  "4 level-1-c/level-2-b/level-3-c/.gitkeep",
  "4 level-1-c/level-2-c/level-3-a/.gitkeep",
  "4 level-1-c/level-2-c/level-3-b/.gitkeep",
  "4 level-1-c/level-2-c/level-3-c/.gitkeep",
]
`;

exports[`should have name property 1`] = `
Array [
  "1 .gitkeep",
  "1 level-1-a",
  "1 level-1-b",
  "1 level-1-c",
  "2 .gitkeep",
  "2 level-2-a",
  "2 level-2-b",
  "2 level-2-c",
  "2 .gitkeep",
  "2 level-2-a",
  "2 level-2-b",
  "2 level-2-c",
  "2 .gitkeep",
  "2 level-2-a",
  "2 level-2-b",
  "2 level-2-c",
  "3 .gitkeep",
  "3 level-3-a",
  "3 level-3-b",
  "3 level-3-c",
  "3 .gitkeep",
  "3 level-3-a",
  "3 level-3-b",
  "3 level-3-c",
  "3 .gitkeep",
  "3 level-3-a",
  "3 level-3-b",
  "3 level-3-c",
  "3 .gitkeep",
  "3 level-3-a",
  "3 level-3-b",
  "3 level-3-c",
  "3 .gitkeep",
  "3 level-3-a",
  "3 level-3-b",
  "3 level-3-c",
  "3 .gitkeep",
  "3 level-3-a",
  "3 level-3-b",
  "3 level-3-c",
  "3 .gitkeep",
  "3 level-3-a",
  "3 level-3-b",
  "3 level-3-c",
  "3 .gitkeep",
  "3 level-3-a",
  "3 level-3-b",
  "3 level-3-c",
  "3 .gitkeep",
  "3 level-3-a",
  "3 level-3-b",
  "3 level-3-c",
  "4 .gitkeep",
  "4 .gitkeep",
  "4 .gitkeep",
  "4 .gitkeep",
  "4 .gitkeep",
  "4 .gitkeep",
  "4 .gitkeep",
  "4 .gitkeep",
  "4 .gitkeep",
  "4 .gitkeep",
  "4 .gitkeep",
  "4 .gitkeep",
  "4 .gitkeep",
  "4 .gitkeep",
  "4 .gitkeep",
  "4 .gitkeep",
  "4 .gitkeep",
  "4 .gitkeep",
  "4 .gitkeep",
  "4 .gitkeep",
  "4 .gitkeep",
  "4 .gitkeep",
  "4 .gitkeep",
  "4 .gitkeep",
  "4 .gitkeep",
  "4 .gitkeep",
  "4 .gitkeep",
]
`;
